File: sdks/typescript/src/skills.ts
/**
 * Skills sub-API. Reached via `client.skills`.
 *
 * Wraps the bridge's GAP-4 skill surface:
 * * `GET  /v1/skills?q=...&min_confidence=...&agent=...&limit=...`
 * * `GET  /v1/skills/stats`
 * * `GET  /v1/skills/:id`
 */

import { doJsonRequest, isObject, runRequest, type RelixClient } from "./client";
import type { ApiResult, Skill, SkillStats, SkillsSearchInput } from "./types";

export class SkillsAPI {
  constructor(private readonly client: RelixClient) {}

  async search(input: SkillsSearchInput = {}): Promise<ApiResult<Skill[]>> {
    return runRequest(async () => {
      const params: Record<string, unknown> = {
        q: input.query,
        agent: input.agent,
        min_confidence: input.minConfidence,
        limit: input.limit,
        peer: input.peer,
      };
      const data = await doJsonRequest(this.client, "GET", "/v1/skills", undefined, params);
      return parseSkills(data);
    });
  }

  async stats(): Promise<ApiResult<SkillStats>> {
    return runRequest(async () => {
      const data = await doJsonRequest(this.client, "GET", "/v1/skills/stats");
      return parseStats(data);
    });
  }

  async get(skillId: string): Promise<ApiResult<Skill>> {
    return runRequest(async () => {
      const data = await doJsonRequest(
        this.client,
        "GET",
        `/v1/skills/${encodeURIComponent(skillId)}`,
      );
      return parseSkillRow(data);
    });
  }
}

function parseSkills(data: unknown): Skill[] {
  let rows: unknown[];
  if (Array.isArray(data)) {
    rows = data;
  } else if (isObject(data)) {
    const r = data["skills"] ?? data["results"];
    rows = Array.isArray(r) ? r : [];
  } else {
    rows = [];
  }
  return rows.map(parseSkillRow);
}

function parseSkillRow(row: unknown): Skill {
  const r = isObject(row) ? row : {};
  const tags = Array.isArray(r["tags"])
    ? (r["tags"] as unknown[]).filter((x): x is string => typeof x === "string")
    : [];
  const steps = Array.isArray(r["steps"])
    ? (r["steps"] as unknown[]).filter((x): x is string => typeof x === "string")
    : undefined;
  return {
    id: typeof r["id"] === "string" ? (r["id"] as string) : "",
    name: typeof r["name"] === "string" ? (r["name"] as string) : "",
    description: typeof r["description"] === "string" ? (r["description"] as string) : "",
    agentId: typeof r["agent_id"] === "string" ? (r["agent_id"] as string) : undefined,
    confidence: typeof r["confidence"] === "number" ? (r["confidence"] as number) : 0,
    usageCount: typeof r["usage_count"] === "number" ? (r["usage_count"] as number) : 0,
    status: typeof r["status"] === "string" ? (r["status"] as string) : "active",
    version: typeof r["version"] === "number" ? (r["version"] as number) : 1,
    tags,
    steps,
  };
}

function parseStats(data: unknown): SkillStats {
  const r = isObject(data) ? data : {};
  return {
    totalSkills: numField(r, "total_skills"),
    activeSkills: numField(r, "active_skills"),
    deprecatedSkills: numField(r, "deprecated_skills"),
    avgConfidence: typeof r["avg_confidence"] === "number" ? (r["avg_confidence"] as number) : 0,
    totalUsage: numField(r, "total_usage"),
  };
}

function numField(r: Record<string, unknown>, key: string): number {
  return typeof r[key] === "number" ? (r[key] as number) : 0;
}

File: sdks/python/relix/skills.py
"""Skills sub-API — search the skill catalogue, read aggregate stats.

Wraps the bridge's GAP-4 skill surface:

* ``GET /v1/skills`` — semantic search across the catalogue
  (filtered by ``q`` / ``min_confidence`` / ``agent`` / ``limit``).
* ``GET /v1/skills/stats`` — aggregate counts.
* ``GET /v1/skills/:id`` — full skill detail.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from pydantic import BaseModel, ConfigDict, Field

if TYPE_CHECKING:
    from .client import RelixClient


class Skill(BaseModel):
    """One row from :meth:`SkillsAPI.search` or :meth:`get`.

    Schema reflects ``crates/relix-runtime/src/nodes/ai/skill_store.rs``'s
    ``Skill`` table. Optional fields are ones that may be ``NULL`` on
    older rows or absent from search responses (vs the full-detail
    ``get``).
    """

    model_config = ConfigDict(extra="allow")

    id: str
    name: str
    description: str = ""
    agent_id: str | None = None
    confidence: float = 0.0
    usage_count: int = 0
    status: str = "active"
    version: int = 1
    tags: list[str] = Field(default_factory=list)
    steps: list[str] | None = None


class SkillStats(BaseModel):
    """Return value of :meth:`SkillsAPI.stats`."""

    model_config = ConfigDict(extra="allow")

    total_skills: int = 0
    active_skills: int = 0
    deprecated_skills: int = 0
    avg_confidence: float = 0.0
    total_usage: int = 0


class SkillsAPI:
    """Skills sub-API. Reached via :attr:`RelixClient.skills`."""

    def __init__(self, client: "RelixClient") -> None:
        self._client = client

    def search(
        self,
        *,
        query: str | None = None,
        agent: str | None = None,
        min_confidence: float | None = None,
        limit: int | None = None,
        peer: str | None = None,
    ) -> list[Skill]:
        """Search the skill catalogue.

        Args:
            query: Free-form natural-language search string.
            agent: Filter to a single agent id.
            min_confidence: Drop skills below this confidence (0.0–1.0).
            limit: Maximum rows.
            peer: Coordinator alias override.
        """
        params = self._build_search_params(query, agent, min_confidence, limit, peer)
        data = self._client._sync_get("/v1/skills", params=params)
        return self._parse_skills(data)

    async def asearch(
        self,
        *,
        query: str | None = None,
        agent: str | None = None,
        min_confidence: float | None = None,
        limit: int | None = None,
        peer: str | None = None,
    ) -> list[Skill]:
        """Async mirror of :meth:`search`."""
        params = self._build_search_params(query, agent, min_confidence, limit, peer)
        data = await self._client._async_get("/v1/skills", params=params)
        return self._parse_skills(data)

    @staticmethod
    def _build_search_params(
        query: str | None,
        agent: str | None,
        min_confidence: float | None,
        limit: int | None,
        peer: str | None,
    ) -> dict[str, Any]:
        return {
            "q": query,
            "agent": agent,
            "min_confidence": min_confidence,
            "limit": limit,
            "peer": peer,
        }

    @staticmethod
    def _parse_skills(data: Any) -> list[Skill]:
        if isinstance(data, list):
            rows: list[Any] = data
        elif isinstance(data, dict):
            rows = data.get("skills") or data.get("results") or []
        else:
            rows = []
        return [Skill.model_validate(r) for r in rows]

    def stats(self) -> SkillStats:
        """Aggregate counts across the skill catalogue."""
        data = self._client._sync_get("/v1/skills/stats")
        return SkillStats.model_validate(data or {})

    async def astats(self) -> SkillStats:
        """Async mirror of :meth:`stats`."""
        data = await self._client._async_get("/v1/skills/stats")
        return SkillStats.model_validate(data or {})

    def get(self, skill_id: str) -> Skill:
        """Full detail for one skill, including step list + version history."""
        data = self._client._sync_get(f"/v1/skills/{skill_id}")
        return Skill.model_validate(data or {})

    async def aget(self, skill_id: str) -> Skill:
        """Async mirror of :meth:`get`."""
        data = await self._client._async_get(f"/v1/skills/{skill_id}")
        return Skill.model_validate(data or {})
